File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.cpp
#include <string>
#include <vector>
#include <queue>
#include <climits>
using namespace std;
int Min = INT_MAX;

void bfs(vector<vector<int>> wire, int n,int x){
    queue<int> q;
    int visit[101] ={0,};
    if(x==0){
        q.push(wire[1][0]);
        q.push(wire[1][1]);
        visit[wire[1][0]]=1;
        visit[wire[1][1]]=1;
    }
    else{
        q.push(wire[0][0]);
        q.push(wire[0][1]);
        visit[wire[0][0]]=1;
        visit[wire[0][1]]=1;
    }
    int cnt= 2;
    while(q.size()!=0){
        int a = q.front();
        q.pop();
        for(int i=0;i<wire.size();i++){
            if(i==x)
                continue;
            if(wire[i][0] == a and !visit[wire[i][1]]){
                visit[wire[i][1]] = 1;
                q.push(wire[i][1]);
                cnt++;
            }
            else if(wire[i][1] == a and !visit[wire[i][0]]){
                visit[wire[i][0]] = 1;
                q.push(wire[i][0]);
                cnt++;
            }
        }
    }
    int a = cnt;
    int b = n-cnt;
    if(a<b)
        swap(a,b);
    Min = min(a-b,Min);
}

int solution(int n, vector<vector<int>> wires) {
    int answer = -1;
    for(int i=0;i<wires.size();i++){
        bfs(wires,n,i);
    }
    answer = Min;
    return answer;
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.io.*;
import java.util.*;
class Solution {
    static int cnt;
    static ArrayList<Integer>[] roads;
    static int[] visit;
    public int solution(int n, int[][] wires) {
        int answer = Integer.MAX_VALUE;
        
        visit = new int[n+1];
        roads = new ArrayList[n+1];
        for(int i = 0; i <= n; i++) roads[i] = new ArrayList<>();
        for(int i = 0; i < wires.length; i++){
            roads[ wires[i][0] ].add(wires[i][1]);
            roads[ wires[i][1] ].add(wires[i][0]);
        }
        
        for(int i = 0; i < wires.length; i++){
            dfs(1, wires[i][0], wires[i][1]); // 시작점, 금지 길
            int x = n - cnt;
            x = Math.abs(x - cnt);
            answer = Math.min(answer, x);
            
            Arrays.fill(visit,0);
            cnt = 0;
        }
        
        
        return answer;
    }
    static void dfs(int s, int x1, int x2){
        if(visit[s] == 1) return;
        cnt++; visit[s] = 1;
        
        for(int e : roads[s]){
            if((s == x1 && e == x2) || (s == x2 && e == x1)) continue; //끊은 길
            dfs(e, x1, x2);
        }
    }
}